### ISSUE TITLE
Allow for more frequent etcd compaction but still long-enough history in etcd

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#80513]
Currently, the etcd compaction algorithm that we use is as following:
- we store "etcd revision" from the moment of last compaction under a special key in etcd
- then periodically even N seconds/minutes we: 
  - compare to see if (local compact time) = (remote compact time)
  - if yes, we increment both local and remote compact time to "now" and do a compaction
  - if not, set local compact time to the remote one

This basically means, that for a given interval, we in etcd we store a history of length roughly between <interval, 2*interval>.

However, this is a significant issue that more frequent compaction means less history in etcd.
So for scalability and performance reasons, we would like to do compaction more frequently, but not necessary reduce amount of history in etcd.

So at the high level, I would like to be able to store a history of length roughly between <(N-1)*interval, N*interval>.

One potential option to do that would be to store last N values in "remote_compact_time", but there are probably better options to do that.

@jpbetz @jingyih @wenjiaswe @mborsz @xiang90

**Comment by wojtek-t** [comment]
Actually, when looking into: https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/maintenance.md, etcd auto-compaction is doing exactly that for N=10 :)

So if we e.g. would like to compact every 30s, we still would have ~5m of history, which would be great.
What I'm afraid of though is what happens if etcd is periodically restarting because of some reason.

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability-feature-requests

**Comment by mborsz** [comment]
> Actually, when looking into: https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/maintenance.md, etcd auto-compaction is doing exactly that for N=10 :)
> 
> So if we e.g. would like to compact every 30s, we still would have ~5m of history, which would be great.
Do you mean setting '--auto-compaction-mode=periodic --auto-compaction-retention=5m'?

IIUC the doc above, the behavior will depend on the exact etcd's version we use:
* in 3.3.0, 3.3.1, 3.3.2, it will indeed do compact every 30 seconds with 5 minutes window, but
* in 3.3.3, it will compact every 5 minutes with 5 minutes window (there is a special case for retention windows shorter than 1 hour).

**Comment by jpbetz** [comment]
> IIUC the doc above, the behavior will depend on the exact etcd's version we use:
> 
> * in 3.3.0, 3.3.1, 3.3.2, it will indeed do compact every 30 seconds with 5 minutes window, but
> * in 3.3.3, it will compact every 5 minutes with 5 minutes window (there is a special case for retention windows shorter than 1 hour).

Yes, since the etcd functionality is version dependent, in the short term I'd prefer we update the kubernetes compactor (https://github.com/kubernetes/kubernetes/blob/1dac5fd14a54ac4972339dbe55f9f03688fd7542/staging/src/k8s.io/apiserver/pkg/storage/etcd3/compact.go#L72).

/cc @jingyih @wenjiaswe

**Comment by ptabor** [comment]
@jpbetz Do you recall why etcd decided to not do x/10 since 3.3.3. ? 
The former behavior seems more aligned with k8s needs and would allow compacting to 10min. 

There is benefit in running  etcd-internal compactor, as in case of problems (e.g. caused by too long time horizont), apiServer might not be able to trigger compaction, while the compaction would enable apiServer to operate.

**Comment by jpbetz** [comment]
> Joe Betz Do you recall why etcd decided to not do x/10 since 3.3.3. ?

No, I think I missed change going through :( I'm in favor of having the ability to compact as often as desired for whatever retention window is needed.

**Comment by serathius** [comment]
Note; since v3.4 etcd supports auto-compaction with proper interval handling as proposed by @wojtek-t

**Comment by wojtek-t** [comment]
I would be supportive for it.

We should bring this up to SIG api-machinery - @deads2k @jpbetz - FYI

**Comment by serathius** [comment]
Would be interested in landing this.

Do we need a full KEP for this? @deads2k @jpbetz

**Comment by jpbetz** [comment]
> Would be interested in landing this.
> 
> 
> 
> Do we need a full KEP for this? @deads2k @jpbetz 

Yes, both the apiserver and etcd are compatible and configured correctly. We'd need to understand how that would be done and have an opportunity for feedback. A KEP seems like the right way to do that.

**Comment by mm4tt** [comment]
+1 for KEP. As discussed, we should cover there things like:

1. Potential regressions. Let's consider a simplest scenario for that: a non-HA deployment (single api-server instance) with etcd deployed separately (e.g on a different VM). As of now, as long as api-server is down the compaction doesn't happen which has some benefits. When api-server gots back up the clients can resume watches without the risk of getting ERR_COMPACTED. If we switch to the etcd's auto-compaction, then compaction will happen when api-server is down and when it goes back up we may hit a thundering herd issue - all watches fail to resume and reLIST which may cause performance degradation. 

2. Rollout of the feature. We should provide a safe way for every cluster deployer to switch from api-server's compact.go to etcd's auto compaction. From a cursory investigation, it seems that there should be no harm in enabling auto-compact when api-server's compaction is still operational and then turning the api-server's compaction off, but we should evaluate this more thoroughly.
